openmdao/core/tests/test_impl_comp.py

```diff
@@ -500,6 +500,120 @@ def guess_nonlinear(self, inputs, outputs, resids):
         for key, val in totals.items():
             assert_near_equal(val['rel error'][0], 0.0, 1e-9)
 
+    def test_guess_nonlinear_residuals(self):
+
+        class ImpWithInitial(om.ImplicitComponent):
+            """
+            An implicit component to solve the quadratic equation: x^2 - 4x + 3
+            (solutions at x=1 and x=3)
+            """
+            def setup(self):
+                self.add_input('a', val=1.)
+                self.add_input('b', val=-4.)
+                self.add_input('c', val=3.)
+
+                self.add_output('x', val=0.)
+
+                self.declare_partials(of='*', wrt='*')
+
+            def apply_nonlinear(self, inputs, outputs, residuals):
+                a = inputs['a']
+                b = inputs['b']
+                c = inputs['c']
+                x = outputs['x']
+                residuals['x'] = a * x ** 2 + b * x + c
+
+            def linearize(self, inputs, outputs, partials):
+                a = inputs['a']
+                b = inputs['b']
+                c = inputs['c']
+                x = outputs['x']
+
+                partials['x', 'a'] = x ** 2
+                partials['x', 'b'] = x
+                partials['x', 'c'] = 1.0
+                partials['x', 'x'] = 2 * a * x + b
+
+            def guess_nonlinear(self, inputs, outputs, resids):
+                # Default initial state of zero for x takes us to x=1 solution.
+                # Here we set it to a value that will take us to the x=3 solution.
+                outputs['x'] = 5.0
```

Comment:
The story was to make sure we had valid residuals when we reached this point inside of `guess_nonlinear`.  You should add an assert in each of these tests right here in the guess_nonlinear function that asserts that the residual 'x' is non-zero.